### PR TITLE
DrivAerML: AdamW beta2 sweep (faster variance adaptation)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

AdamW's default beta2=0.999 means the second-moment estimate (squared gradient running average) has a half-life of ~693 steps. Over 467 epochs with ~394 steps/epoch (~184k total steps), stale gradient variance estimates can accumulate, especially through the cosine LR peak/trough transitions.

Reducing beta2 to 0.99 (half-life ~69 steps) or 0.95 (half-life ~14 steps) makes the optimizer adapt faster to gradient magnitude changes. This is particularly relevant for DrivAerML where gradient norms vary dramatically across cosine phases — the optimizer should scale its steps based on RECENT gradient variance, not stale estimates from hundreds of steps ago.

This approach targets a different mechanism than warmup/RAdam (which only help early training) or schedule changes (which change the LR shape). Beta2 affects how the optimizer responds to gradient changes THROUGHOUT training.

Reference: "Decoupled Weight Decay Regularization" (Loshchilov & Hutter 2019) and the observation that lower beta2 often helps in settings with non-stationary gradient statistics.

## Instructions

Add a --adam-beta2 flag to train.py. Run two values:

**Run 1: beta2=0.99**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --adam-beta2 0.99 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-adamw-beta2
```

**Run 2: beta2=0.95**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --adam-beta2 0.95 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group dm-adamw-beta2
```

Implementation: pass beta2 to AdamW as `betas=(0.9, args.adam_beta2)`. Keep beta1=0.9 (default). No --grad-clip, no --weight-decay. Report best val_primary/surface_rel_l2_pct for both runs.

## Baseline

- **val_primary/surface_rel_l2_pct:** 3.997% at epoch 467
- **Config:** 4L/512d/8H, AdamW lr=5e-4 (betas=(0.9, 0.999) default), T_max=30, no-EMA, Fourier
- **W&B run:** bht6h42t
- **Reproduce:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT megumi:

## Results

### Run 1: beta2=0.99 (W&B: 95teq8ui)
- **Best val_primary/surface_rel_l2_pct:** 7.11% at epoch 140 (vs baseline 3.997%)
- **Current status (epoch 165):** Recovering from 3rd instability spike, val ~18%
- **Peak memory:** ~6 GB
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --adam-beta2 0.99 \
  --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 \
  --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent megumi --wandb-name "megumi/dm-beta2-0.99" --wandb-group dm-adamw-beta2
```

### Run 2: beta2=0.95 (W&B: 4dladx7q)
- **Best val_primary/surface_rel_l2_pct:** 13.33% at epoch 40 (vs baseline 3.997%)
- **Current status (epoch 165):** Permanently diverged — grad_norm routinely 100-57,000+, val oscillating 15-40%
- **Peak memory:** ~6 GB
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --adam-beta2 0.95 \
  --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 \
  --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent megumi --wandb-name "megumi/dm-beta2-0.95" --wandb-group dm-adamw-beta2
```

### Comparison vs baseline (3.997% at epoch 467, beta2=0.999)

| Metric | beta2=0.999 (baseline) | beta2=0.99 | beta2=0.95 |
|--------|----------------------|------------|------------|
| Best val surface_rel_l2_pct | **3.997%** | 7.11% | 13.33% |
| Best epoch | 467 | 140 | 40 |
| Training stability | Stable to 467+ | Spikes at ep76, 122, 149, 156 | Diverged from ep47 |
| Grad norm (stable) | ~0.4-0.7 | ~0.4-0.7 (between spikes) | N/A (chronically elevated) |

### What happened

**The hypothesis was partially confirmed in early training but both variants fail in long training:**

1. **Early training (ep 1-45):** beta2=0.95 showed clear advantages — faster convergence, smaller spikes at cosine LR transitions, better handling of gradient magnitude changes. The hypothesis mechanism (faster variance adaptation at LR transitions) was visible and real.

2. **Mid training (ep 45-75):** beta2=0.95 began diverging (ep47 onward), with exponentially growing grad_norms. beta2=0.99 remained stable and pulled ahead, reaching 9.16% by ep70.

3. **Late training (ep 76+):** beta2=0.99 started showing periodic instability spikes (ep76, 122, 149, 156) with increasingly severe grad_norm explosions (10→26→138→35). It recovers each time but the spikes are getting worse and more frequent. Best metric was achieved at ep140 (7.11%) but the run cannot sustain stable training long enough to match the baseline's ep467 performance.

4. **Root cause:** Lower beta2 makes the second-moment estimate more responsive to recent gradient variance, but this cuts both ways. When a batch produces an unusually large gradient, the second-moment adapts too quickly, the effective step size spikes, and this cascades into instability. Without gradient clipping to cap these excursions, the faster adaptation becomes a liability in long training.

5. **Neither variant beats the baseline.** The default beta2=0.999 achieves 3.997% at epoch 467 precisely because its slow adaptation provides a stabilizing "memory" effect that smooths over gradient spikes across hundreds of epochs.

### Suggested follow-ups

1. **beta2=0.99 + gradient clipping:** The 0.99 variant showed promising convergence speed between spikes. Pairing it with --grad-clip 1.0 or 2.0 could eliminate the instability while preserving the faster adaptation benefits. This is the most promising follow-up.

2. **beta2=0.995:** A middle ground between 0.999 and 0.99 that might balance adaptation speed with stability. Half-life ~138 steps vs 693 (0.999) or 69 (0.99).

3. **Adaptive beta2 schedule:** Start with lower beta2 (0.99) in early training when fast adaptation helps, then anneal toward 0.999 for stability in late training.

Note: Both runs are still running (epoch ~165 of 999, ~115 min remaining in timeout). Results will continue to stream to W&B but the instability pattern is well-established.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3128 because ADVISOR: Closing — AdamW beta2 sweep is a dead end on DrivAerML. beta2=0.99: 7.113% best (+78% vs baseline). beta2=0.95: 13.332% best (+234%), grad norms reached 1.26M. The long second-moment memory of beta2=0.999 is a stability feature — faster adaptation amplifies outlier batches into cascading gradient spikes at cosine LR peaks. Clear monotonic relationship: 0.999 > 0.99 > 0.95. Leave beta2 at default. Good mechanistic analysis, megumi..
